README.md


# Responsive Widgets

[![pub package](https://img.shields.io/pub/v/responsive_widgets.svg)](https://pub.dev/packages/responsive_widgets)

This plugin helps to create responsive widgets, that makes auto-size with the **proportion between reference screen size** 
(width, height)  with the screen that the app is running. The package only changed the original widgets, like "Container" 
to apply a function that make this calculation.

## Important notes
Since the version 2.0.0, we started to use the package ScreenUtils ([flutter_screenutils](https://pub.dev/packages/flutter_screenutil))
to calculate all, have so many breaking changes, so if you used this package on the past, you will need **stop** using 
`ResponsiveWidgets.getSize()` and use the ScreenUtils that is inside of this package. For more informations about the 
use of this package, view the original docs of the package: 
[https://pub.dev/packages/flutter_screenutil](https://pub.dev/packages/flutter_screenutil).<br>
Basically now this package turned in a helper of this other package, helping to use more fast in the widgets and helping 
with the web version using the LayoutBuilder automatically when used `ResponsiveWidgets.builder()`<br><br>
If you want use the previous version of the package, use the version `1.1.0` and in your pubspec.yaml: `responsive_widgets: 1.1.0`
<br><br>



## Features
| Feature / Working with                 | Progress |
| :------------------------------------- | :------: |
| Android                                |    ✅    |
| iOS                                    |    ✅    |
| Smartphones                            |    ✅    |
| 5.5 inch screens                       |    ✅    |
| Vertical mode                          |    ✅    |
| Horizontal mode                        |    ✅    |
| Tablet                                 |    ✅    |
| Tablet on vertical mode                |    ✅    |
| Tablet on horizontal mode              |    ✅    |
| Support for web                        |    ✅    |


## Getting Started  

First of all you need to use the code `ResponsiveWidgets.init(context)` to make the plugin works. 
Use `ResponsiveWidgets.builder(child: )`, to recalculate the proportion every time the sizes of device changes. 
Inclusive when the device turned to left/right... The code need to be placed on the first screen of the app, inside of
the build method, like this: [See also in code](https://github.com/LucazzP/responsive_widgets/blob/master/example/lib/home_screen.dart)


    class  _HomeScreenState  extends  State<HomeScreen> {
	    @override
	    Widget  build(BuildContext context) {
	    
		    ResponsiveWidgets.init(context,
			    height: 1920, // Optional
			    width: 1080, // Optional
				allowFontScaling: true, // Optional
		    );
		    
		    return  ResponsiveWidgets.builder(
				height: 1920, // Optional
			    width: 1080, // Optional
				allowFontScaling: true, // Optional
		        child: Scaffold(
                        body: Container()
                    ),
		    );
		}
	}


## Widgets  

#### ContainerResponsive

    ContainerResponsive(
	    child: Widget,
	    height: double, // Responsive height
	    heightResponsive: bool, // Enable/Disable Responsive height
	    width: double, // Responsive width
	    widthResponsive: , bool// Enable/Disable Responsive width
    )

  

#### SizedBoxResponsive

    SizedBoxResponsive(
	    child: Widget,
	    height: double, // Responsive height
	    width: double, // Responsive width
    )



#### EdgeInsetsResponsive (Can be used in any widget with parameter padding)

    Padding(
	    child: Widget,
	    padding: EdgeInsetsResponsive.all(50), // EdgeInsets Responsive padding
    )
  


#### TextResponsive

    TextResponsive(
	    text // Responsive fontSize
    )



#### RaisedButtonResponsive

    RaisedButtonResponsive(
	    child: Widget,
	    onPressed: (){}
    )
  

### Function to recalculate size responsibly

Have special cases that doesn't have Responsive Widgets created, one of this cases is the Positioned, to solve this problem is simple use this function that will return the correct value to the screen.

---
------- Copyed from [flutter_screenutils](https://pub.dev/packages/flutter_screenutil) docs. --------

Pass the px size of the design draft：

Adapted to screen width: `ScreenUtil().setWidth(540)`,

Adapted to screen height: `ScreenUtil().setHeight(200)`,

If your dart sdk>=2.6, you can use extension functions:

example:

instead of :
```
Container(
width: ScreenUtil().setWidth(50),
height:ScreenUtil().setHeight(200),
)
```
you can use it like this:
```
Container(
width: 50.w,
height:200.h
)
```

#### Adapter font:
``` dart
//Incoming font size，the unit is pixel, fonts will not scale to respect Text Size accessibility settings
//(AllowallowFontScaling when initializing ScreenUtil)
ScreenUtil().setSp(28)    
     
//Incoming font size，the unit is pixel，fonts will scale to respect Text Size accessibility settings
//(If somewhere does not follow the global allowFontScaling setting)
ScreenUtil().setSp(24, allowFontScalingSelf: true)

//for example:

Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                Text(
                    'My font size is 24px on the design draft and will not change with the system.',
                    style: TextStyle(
                      color: Colors.black,
                      fontSize: ScreenUtil().setSp(24),
                    )),
                Text(
                    'My font size is 24px on the design draft and will change with the system.',
                    style: TextStyle(
                        color: Colors.black,
                        fontSize: ScreenUtil()
                            .setSp(24, allowFontScalingSelf: true))),
              ],
            )
```

------- End of copy --------

----
  

## Examples


### Mixed, responsive and normal, respectively
![Mixed, responsive and normal, respectively](https://i.imgur.com/96pdtzR.jpg)

### Not responsive page
![Not responsive page](https://i.imgur.com/WR7UQAB.jpg)

### Responsive page
![Responsive page](https://i.imgur.com/3rzKnfW.jpg)


CHANGELOG.md
## 2.0.1
- SizedBoxResponsive fix

## 2.0.0+1
- Changed README.md

## 2.0.0
- Improved all Widgets and removed all unnecessary codes
- Added SizedBoxResponsive
- Breaking changes:
    - Changed the calculation of responsiveness to use the [flutter_screenutils](https://pub.dev/packages/flutter_screenutil).
    - Removed ResponsiveWidgets.getSize()
    - IconResponsive and IconButtonResponsive excluded, because Screen utils does not support(Will be reviewed this question)


## 1.1.0
- Breaking change:
    Fixed imports to be only one: `import 'package:responsive_widgets/responsive_widgets.dart';`, all other imports 
    won't work more

## 1.0.2
- Fixed to run only with context as argument to init.

## 1.0.1
- Added support for Tablet, horizontal mode, and for web.
- To works the new functionalities is necessary to use `ResponsiveWidgets.builder(child: )`, to recalculate the proportion
every time the sizes of device changes. Inclusive when the device turned to left/right...
- `ResponsiveWidgets().init(context)` changed to `ResponsiveWidgets.init(context)`
- Changed type of Flutter project from plugin to package.

(Breaking changes)

## 0.1.10
- Added assert if the plugin is used without init

## 0.1.9
- Changed to static methods on main class

## 0.1.8
- Keys is now Working
- Classes code reformated

## 0.1.7
- Updated Kotlin and Gradle

## 0.1.6
- Updated packages

## 0.1.5
- Ios fixed.

## 0.1.4
- Readme updated.
- Repository url fixed.

## 0.1.3
- Readme updated.

## 0.1.2
- Readme updated.
- Changed getSizeByDp() to getSize().

## 0.1.1
- Changed description.

## 0.1.0
- This plugin helps to create responsive widgets, that makes a auto-size with the proportion between reference Screen Size and the screen that the app is running. The package only changed the original widgets, like "Container" to apply a function that make this calculation.

example/lib/home_screen.dart
import 'package:flutter/material.dart';
import 'package:responsive_widgets/responsive_widgets.dart';
import 'non_responsive_screen.dart';
import 'responsive_screen.dart';

class HomeScreen extends StatefulWidget {
  HomeScreen({Key key}) : super(key: key);

  _HomeScreenState createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  @override
  Widget build(BuildContext context) {
    ResponsiveWidgets.init(
      context,
      height: 1920,
      width: 1080,
      allowFontScaling: false,
    );

    return ResponsiveWidgets.builder(
      height: 1920,
      width: 1080,
      allowFontScaling: false,
      child: Scaffold(
        appBar: AppBar(
          title: Text('Plugin example app'),
          centerTitle: true,
        ),
        body: SingleChildScrollView(
          child: Container(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: <Widget>[
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: <Widget>[
                    RaisedButton(
                      onPressed: () => Navigator.push(
                          context,
                          MaterialPageRoute(
                              builder: (context) => NonResponsiveScreen())),
                      child: TextResponsive("Not Responsive Page"),
                      key: Key('RaisedButton'),
                    ),
                    RaisedButtonResponsive(
                      onPressed: () => Navigator.push(
                          context,
                          MaterialPageRoute(
                              builder: (context) => ResponsiveScreen())),
                      child: TextResponsive("Responsive Page"),
                      key: Key('RaisedButtonResponsive'),
                    )
                  ],
                ),
                TextResponsive(
                  'Plugin example app',
                  key: Key('TextResponsive'),
                ),
                Text(
                  'Plugin example app',
                  key: Key('Text'),
                ),
                ContainerResponsive(
                  key: Key('ContainerResponsive'),
                  height: 100,
                  width: 200,
                  color: Colors.green,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Container(
                      height: 200,
                      width: 100,
                      color: Colors.green,
                    ),
                    SizedBox(
                      width: 20,
                    ),
                    ContainerResponsive(
                      height: 200,
                      width: 100,
                      color: Colors.green,
                    ),
                  ],
                ),
                Container(
                  key: Key('Container'),
                  height: 100,
                  width: 200,
                  color: Colors.green,
                ),
                ContainerResponsive(
                  key: Key('ContainerResponsive2'),
                  height: 100,
                  width: 200,
                  color: Colors.green,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
